File: 0509-fibonacci-number/0509-fibonacci-number.java
class Solution {
    public int fib(int n) {
        int[] dp= new int[n+1];
        return f(n, dp);
    }
    
    public int f(int n, int[] dp){
        if(n<=1) return n;
        if(dp[n]!=0) return dp[n];
        int ans= f(n-1, dp)+f(n-2, dp);
        dp[n]=ans;
        return ans;
    }
}

File: 0509-fibonacci-number/0509-fibonacci-number.cpp
class Solution {
public:
    int fibo(int n, vector<int> & dp){
        if(n<=1) return n;
        if(dp[n]!=-1) return dp[n];
        
        dp[n]= fibo(n-1,dp)+fibo(n-2, dp);
        return dp[n];
    }
    int fib(int n) {
        vector<int> dp(n+1);
        for(int i=0;i<=n;i++){
            dp[i]=-1;
        }
        return fibo(n, dp);
    }
};